### ISSUE TITLE
Incompatibility with paranoia

### ISSUE THREAD

**Issue by dkniffin** [collectiveidea/awesome_nested_set#343]
I believe this gem and [paranoia](https://github.com/rubysherpas/paranoia) are incompatible with each other.

Paranoia enables soft-deletion in a rails app. The way it works is by adding a new column, `deleted_at` which defaults to nil, and setting up `default_scope` that does `where(deleted_at: nil)`. Then when an object is deleted, it really gets `deleted_at` set to the current time.

This problem showed up for me because I was doing this `item.ancestors`, and it was including deleted items in the ancestry tree for the item, even though they were completely unrelated.

I dug into it a bit, and I believe I understand the problem, but I'm not sure what a good fix would be. What's happening is when an item is deleted, awesome_nested_set updates the `lft` and `rgt` columns of the rest of the records, to fill in the gap created by the deleted record. That would be fine in-and-of itself, but [self_and_ancestors](https://github.com/collectiveidea/awesome_nested_set/blob/75178058653925f5569fb19a51a61ba68c3b3095/lib/awesome_nested_set/model/relatable.rb#L13) is looking for all records that are containing or equal to the item, and doesn't respect paranoia's `deleted_at` field, because it's doing an arel query, which bypasses rails' `default_scope`.

A few possible ways to solve this:
- Check if paranoia is being used, perhaps with something like `respond_to?(:deleted_at?)`, and if it is, filter out the deleted items. This might work, but it'd be a bit complex, I think, and it feels like a hack of a solution.
- Don't use Arel to do the queries. I believe this would be the best solution, because it would not only fix incompatibilities with `paranoia`, but also any other usages of `default_scope`. However, I'm not sure if that has wider consequences
- Change the `ancestors` method so that instead of doing `without_self self_and_ancestors`, it does something more like what `self_and_ancestors` is doing internally, but with `lt` and `gt` instead of `lteq` and `gteq`

For now, I've found a workaround that works decently enough, which is to do `ancestors.reject(&:deleted_at)`

**TL;DR** Paranoia and acts_as_nested_set are currently incompatible with each other.

**Comment by hui-green** [comment]
I also have this problem,  I add deleted_at to scope, it works if I don't restore the node, after restore, the whole tree will be in a mass, so we also need to consider the restore thing

**Comment by rhnasc** [comment]
That impacts also the depth attribute, which lead up to some inconsistencies on our applications... Acknowledging the existence of soft-deleted cases when performing the `self_and_ancestors` (your second bullet basically) looks good to me. Ignoring soft-deleted records wherever they may be retrieved by the gem.

**Comment by denodster** [comment]
this is my workaround for this particular problem (we have our own home rolled soft delete mechanism)

```
  def soft_delete
    if self.boss_id
      self.deleted_at_boss_id = self.boss_id
      self.boss_id = nil #this forces awesome_nested_set to behave
      self.save(validate: false)
    end
    self.deleted_at = DateTime.now
    self.save(validate: false)
  end

  def restore_soft_delete
    self.deleted_at = nil
    self.save(validate: false)
    if self.deleted_at_boss_id
      self.boss_id = self.deleted_at_boss_id
      self.deleted_at_boss_id = nil
      self.save(validate: false)
    end
  end
```
Basically we have a seperate field that we move the parent_id into, which causes nested set to resolve itself before we do the delete, likewise when we restore, we restore first and then fix the parent_id.
